### DISCUSSION TITLE
How to import MinimalLib into IPyWidget?

### DISCUSSION THREAD

**Discussion by heberleh** [rdkit/rdkit#4862] [Q&A] [answered]
I'm trying to use the MinimalLib within a Jupyter plugin, but in this case, I can't import the library using the standard way (import using <script></script> in the index.html) because I don't have an index.html.

I tried this, which in theory would not continue running the code until the rdkit library is loaded in `window`.

```javascript
(async () => {
  await import(/*webpackIgnore: true*/ './rdkit/RDKit_minimal');
  // await require(/*webpackIgnore: true*/ './rdkit/RDKit_minimal.js');
})();
```
but webpack doesn't like it. It tries to compile the files.

I tried to use "exclude" and IgnorePlugin in the webpack config too, unsuccessfully.

I'm using the cookicutter for ipywidgets in Typescript: https://github.com/jupyter-widgets/widget-ts-cookiecutter

This is related to #4848 

Update Workaround:

Because the webpack and jupyter-plugin scripts were not making the files accessible in the Front-end,
I tried to use the URL to the library instead, and it worked, but I don't consider this an answer to the question.

```javascript
import * as $ from 'jquery';
$.getScript(
  'https://unpkg.com/@rdkit/rdkit/Code/MinimalLib/dist/RDKit_minimal.js',
  (data: any, textStatus: any, jqxhr: any) => {
    if (window.initRDKitModule) {
      window.initRDKitModule().then((RDKit: any) => (window.RDKit = RDKit));
    }
  }
);
```

**Comment by MichelML** [comment, marked as answer]
I'll try to give a better answer later @heberleh but if you're workaround is somewhat stable you can use it and do it in plain javascript like so https://stackoverflow.com/a/14521482, tldr:

```javascript
const script = document.createElement("script");
script.onload = function () {
    if (window.initRDKitModule) {
      window.initRDKitModule().then((RDKit: any) => (window.RDKit = RDKit));
    }
};
script.src = "https://unpkg.com/@rdkit/rdkit@2021.9.3/Code/MinimalLib/dist/RDKit_minimal.js";
document.head.appendChild(script);
```

Notice I have specified the package version, which is something you might want to do to guarantee consistency for a given release of your plugin.

To make it a bit cleaner, I even think you could write it like this:

```javascript
const script = document.createElement("script");
script.onload = function () {
    if (window.initRDKitModule) {
      window.initRDKitModule().then((RDKit: any) => (window.RDKit = RDKit));
    }
};
script.src = "https://unpkg.com/@rdkit/rdkit@2021.9.3"; // <-- shorter url
document.head.appendChild(script);
```

replacing `https://unpkg.com/@rdkit/rdkit@2021.9.3/Code/MinimalLib/dist/RDKit_minimal.js` for `https://unpkg.com/@rdkit/rdkit@2021.9.3` which points to the former.

--------------------------------

To help scope the problem, a stack trace of the error(s) you're getting with webpack would help - but disclaimer I ran into problems with webpack a while ago (I only remember vaguely what they were about though) thus why I wrote the readme of https://www.npmjs.com/package/@rdkit/rdkit loading the script in a script tag instead.

If you're plugin is currently open source on github I can also have a look.

**Reply by heberleh** [reply]
Thank you so much, @MichelML!
Solution 1 (long url) worked (but 2nd didn't)! So I can remove the jquery from the project, that is great.

I will stick to this solution for now. The code is not public — yet. I need to start the “paperwork” and get approval.
Once it is public, I may come back to the questions I made here to point to my code.

**Reply by MichelML** [reply]
Glad the approach worked for you! Best of luck for the approval :pray:

**Comment by MichelML** [comment]
@heberleh side note, you seem to use rdkit minimal lib for js fairly extensively, if you have any idea on how to improve it, can you log some issues here https://github.com/MichelML/rdkit-js

hopefully this will become the official repo by the end of jan-february

see discussion here https://github.com/rdkit/rdkit/discussions/4124

**Reply by heberleh** [reply]
I will do so! But I don't see a Discussions tab there, so far I had mostly questions and overall the library met all my needs except the lack of a `mol.getCoordinates(drawOpts)`. I took me some time to decode the paths from the svg to get the x,y from atoms.
In the SVG, but maybe this is from RDKit and not RDKit.js, the elements don't have classes that differentiate them, other than the "commands" (attribute d). For example, a double bond1 can have 4 paths (==) and all of them have the same classList. I cannot do a search for: bond1, first bond (--), part1 (-), and then part 2 (-) of this 2-parts-edge (--), to get x,y from source and then target. In rings, == is one long and one short, and to get x,y I need to find the longest one (or believe it will be always the first in the svg).

I can write the info in the other repo if you want. It may make more sense when I have my repo public, so I can connect to examples of codes/applications.

Let me know where I should bring questions and suggestions.
Many thanks again for the help!

**Reply by DavidACosgrove** [reply]
Hi @heberleh, as it happens I am working on the code that generates SVGs at the moment, so perhaps some more could be added that might help.  Please let me know exactly what you'd like and I'll discuss it with Greg and see whether we can implement it.  In the meantime, there is the drawOption splitBonds which, if set True, will always draw the bonds in two halves, and then uses the class 'bond-1 atom-0' for one end 'bond-1 atom-1' at the other, for example.  You will still have the issue of multiple lines with the same class with double and triple bonds.

**Reply by greglandrum** [reply]
@heberleh : I don't think @MichelML is actually planning on doing his own version of MinimalLib. Based on the previous discussions the plan there is to have an official repo which handles the build and packaging, but it will still be using RDKit code for the library itself.

So this remains a reasonable place to discuss what is/isn't in MinimalLib and how to use the functionality.

@MichelML please correct me if I'm wrong about that.

**Reply by MichelML** [reply]
@greglandrum I agree this is a reasonable place, and sorry to have maybe push too quickly on logging stuff over the new repo.

What I think might happen though is people will actually log js-minimallib related things in the new repo, and this will get ambiguous and not sure what's the best path forward.

We'll have to be clear that the new repo is for js examples and additional utilities "on top of the minimallib", but not the minimallib itself, but even then, I wouldn't be surprised if active users of the minimallib in a js context start logging minimal lib request in the js repo. Furthermore, I think that walkthroughs such as using the rdkit js minimallib with webpack (current discussion) will belong to the new repo (or at least partly). Anyhow, I'm sure you get the ambiguity, I suggest we rediscuss this when the new repo is ready.

**Reply by heberleh** [reply]
> Hi @heberleh, as it happens I am working on the code that generates SVGs at the moment, so perhaps some more could be added that might help. Please let me know exactly what you'd like and I'll discuss it with Greg and see whether we can implement it. In the meantime, there is the drawOption splitBonds which, if set True, will always draw the bonds in two halves, and then uses the class 'bond-1 atom-0' for one end 'bond-1 atom-1' at the other, for example. You will still have the issue of multiple lines with the same class with double and triple bonds.

So I think that what is missing is just a way to find "unique" elements so that I know what I am selecting. This can help future users who try to hack the svg.

With the splitBonds, it almost solve the problem, but some persist.

In my case, I try to find the x,y of atoms.

```html
<path class="bond-8 atom-8" d="M 207.1,67.6 L 191.9,81.3" style="fill:none;fill-rule:evenodd;stroke:#000000;stroke-width:2.0px;stroke-linecap:butt;stroke-linejoin:miter;stroke-opacity:1"></path>
```
- here, to get atom-8 coordinates, the SVG don't tell me if I should use the `M x y` or the `L x y`. I would assume the to use M for the first path and L for the latest path, but I don't know if this would be always a guaranteed order.

Let's imagine a ring, with 2 edges (bond-8), one is shorter than the other. Now I have 2 edges for bond-8. To get the x,y for the atom, I need to calculate which one is the longest, then I use its x,y's for the source and target.

Now imagine a double or triple bond, these are easier with the "splitBoinds". In this case, I select all of them and get the average(x) and average(y). However, I still need to check if their length are similar, so I can check if they are in a ring or not.

In both cases, I think I still need to calculate where the paths connect, so I know where are the source/target x,y (left or right?).

In summary, I think that if the parser already has information about the molecule when it is creating the SVG, that info could be encoded in the svg.

Examples:
- where is source/target/atom? right? left? or just the x,y of the source/target directly encoded.
- < g> groups can help to separate groups of paths
- isPartOfRing may help to deal with that special case < g isPartofRing=true>< path>....</g>
- isShortEdgeFromRing could be another one to indicate the short edges in the rings
- < g nBond=N> where N is 1, 2, 3... can help to know if that group of paths represent single double bonds etc.
- for the Text-atoms (N, C, Cl, NH, etc), I am calculating the convex hull of all points in the "commands" (path.d) and getting the avg(x), and avg(y)

For my case, I think it is working well... the logics that I use to extract x,y, but I am uncertain if it will always work.
I have just used a bigger test set, and it failed. I will investigate if the problem is the regex I'm using to identify the atoms from the smiles string or the x,y extractor.

**Reply by greglandrum** [reply]
> @greglandrum I agree this is a reasonable place, and sorry to have maybe push too quickly on logging stuff over the new repo.
> 
> What I think might happen though is people will actually log js-minimallib related things in the new repo, and this will get ambiguous and not sure what's the best path forward.
> 
> We'll have to be clear that the new repo is for js examples and additional utilities "on top of the minimallib", but not the minimallib itself, but even then, I wouldn't be surprised if active users of the minimallib in a js context start logging minimal lib request in the js repo. Furthermore, I think that walkthroughs such as using the rdkit js minimallib with webpack (current discussion) will belong to the new repo (or at least partly). Anyhow, I'm sure you get the ambiguity, I suggest we rediscuss this when the new repo is ready.

That all sounds great, and we can always just reference github issues from one repo in the other (or if the new repo ends up as part of the rdkit organization we can directly transfer them).

**Reply by greglandrum** [reply]
@heberleh I'm not 100% sure what you're trying to do, but it sounds like the output of the TagAtoms() method of the SVG renderer could be useful to you. This isn't currently accessible from MinimalLib, but you can try it out in Python.
Here's what the output looks like for CCO:
```
<path  d='M 11.4,124.4 L 119.1,62.2' class='bond-selector bond-0 atom-0 atom-1' style='fill:#fff;stroke:#fff;stroke-width:4.0px;fill-opacity:0;stroke-opacity:0' />
<path  d='M 119.1,62.2 L 226.8,124.4' class='bond-selector bond-1 atom-1 atom-2' style='fill:#fff;stroke:#fff;stroke-width:4.0px;fill-opacity:0;stroke-opacity:0' />
<circle  cx='11.4' cy='124.4' r='16.6' class='atom-selector atom-0' style='fill:#fff;stroke:#fff;stroke-width:1px;fill-opacity:0;stroke-opacity:0' />
<circle  cx='119.1' cy='62.2' r='16.6' class='atom-selector atom-1' style='fill:#fff;stroke:#fff;stroke-width:1px;fill-opacity:0;stroke-opacity:0' />
<circle  cx='226.8' cy='124.4' r='16.6' class='atom-selector atom-2' style='fill:#fff;stroke:#fff;stroke-width:1px;fill-opacity:0;stroke-opacity:0' />
```
That gives you a clear set of entities which can be selected to get the individual atoms and bonds.

You use it in Python like this:
```
d2d = Draw.MolDraw2DSVG(250,200)
d2d.DrawMolecule(tm)
d2d.TagAtoms(tm)
d2d.FinishDrawing()
txt = d2d.GetDrawingText()
```

**Reply by greglandrum** [reply]
I blogged about this and showed a crude PoC of something built on top of it here: https://rdkit.blogspot.com/2019/08/an-interactive-rdkit-widget-for-jupyter.html

**Reply by heberleh** [reply]
@greglandrum  Thanks for the input!
I was just trying to get x,y for each atom and I think I completed it successfully now. I tested with a bigger set of molecules, and it seems to work.
I also was able to set the color of atoms to black, which is perfect for my application.

**So, basically, now I will set the RDKit drawer as default for my application, instead of using the smilesDrawer.**

This is of great relevance since the users will now be able to use it in combination with Python RDKit, publishing the images from RDKit or my application all in the same standards.

Soon it should be public, and it is about XAI as you can imagine :)

Thank you so much for all the help during the last days, @greglandrum @ptosco @MichelML @DavidACosgrove.
Please let me know if I can thank you (for the technical help to set up RDKit MinimalLib) in the Acknowledgments section of our future publication (potentially BMC but may be other; we need permission to cite names there :).

**Reply by greglandrum** [reply]
@heberleh awesome. Congrats!
Definitely let us know when there's something out that we can see/play with.
and, yeah, I'm always happy to be acknowledged. ;-)

**Reply by DavidACosgrove** [reply]
second all of that

**Reply by MichelML** [reply]
@heberleh I give my permission to be acknowledged too, it'd be nice and appreciated :ok_hand: (nice of you to offer by the way)

**Reply by ptosco** [reply]
@heberleh I second all of the above too, happy to be acknowledged, many thanks for offering!

**Comment by ptosco** [comment]
@heberleh I second all of the above too, happy to be acknowledged, many thanks for offering!
